Utils/src/GoodJetsProducer.cc

```diff
@@ -133,122 +136,113 @@ GoodJetsProducer::~GoodJetsProducer()
-           if(!good && !TagMode_) return false;
-           result &= good;
-        }
-      }
-   }
-   // put in the event
-   iEvent.put(std::move(prodJets));
-   iEvent.put(std::move(jetsMask),"JetIDMask");
-   iEvent.put(std::move(leptonMask),"JetLeptonMask");
-   auto passing = std::make_unique<bool>(result);
-   iEvent.put(std::move(passing),"JetID");
-   return true;
-   
+GoodJetsProducer::filter(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+	using namespace edm;
+	// load to be excluded leptons, isotracks and photons
+	std::vector<edm::Handle<edm::View<reco::Candidate>>> excludeHandles;
+	excludeHandles.reserve(SkipTag_.size());
+	if(ExcludeLeptonIsoTrackPhotons_ && !SkipTag_.empty()) {
+		// loop over each edm::InputTag
+		for (unsigned iC = 0; iC < SkipTag_.size(); ++iC) {
+			// get each collection the edm::InputTag corresponds to
+			edm::Handle<edm::View<reco::Candidate>> cleanCands;
+			iEvent.getByToken(SkipTok_[iC], cleanCands);
+			if (cleanCands.isValid()) {
+				excludeHandles.push_back(cleanCands);
+			}
+			else {
+				edm::LogWarning("TreeMaker")<<"Warning: skip tag not valid in GoodJetsProducer: "<<SkipTag_[iC];
+			}
+		}
+	}
+
+	auto prodJets = std::make_unique<std::vector<Jet>>();
+	auto jetsMask = std::make_unique<std::vector<bool>>();
+	auto leptonMask = std::make_unique<std::vector<bool>>();
+	bool result=true;
+	edm::Handle< edm::View<Jet> > Jets;
+	iEvent.getByToken(JetTok_,Jets);
+	if(Jets.isValid()) {
+		prodJets->reserve(Jets->size());
+		jetsMask->reserve(Jets->size());
+		leptonMask->reserve(Jets->size());
+		for(const auto & iJet : *Jets) {
+			if (maxEta_>0 and std::abs(iJet.eta())>maxEta_) continue;
+			if (!saveAllPt_ && ( (!invertJetPtFilter_ && iJet.pt() <= jetPtFilter_) || (invertJetPtFilter_ && iJet.pt() > jetPtFilter_) ) ) continue;
+			if (!iJet.isPFJet()) continue;
+			//check for a candidate with infinite energy, drop the associated jet
+			if (iJet.numberOfDaughters()>0 and std::isinf(iJet.daughterPtr(0)->energy())) continue;
+
+			bool skip=false;
+			if (ExcludeLeptonIsoTrackPhotons_ && !excludeHandles.empty()) {
+				for (const auto & excludeHandle : excludeHandles) {
+					for (const auto & exclude : *excludeHandle) {
+						if (std::abs(iJet.pt() - exclude.pt() ) / exclude.pt() < 1 && deltaR(iJet.p4(),exclude.p4()) < JetConeSize_ ) {
+							skip=true;
+							break;
+						}
+						if (skip) break; //no need to keep checking
+					}
+				}
+				if (skip) {
+					prodJets->push_back(Jet(iJet));
+					jetsMask->push_back(true);
+					leptonMask->push_back(true);
+					continue;
+				}
+				else leptonMask->push_back(false);
+			}
+			else leptonMask->push_back(false);
+
+			//calculate the PFJetID decision
+			bool good = true;
+			double varval = 0.0;
+			for(unsigned v=0; v<varnames_.size(); ++v) {
+				if(varnames_[v]=="neutralHadronEnergyFraction" || varnames_[v]=="nhf")      varval = iJet.neutralHadronEnergyFraction();//gives raw energy in the denominator
```

Comment:
make the available var names into enums to avoid N*M string comparisons
(or go for the full factory approach of JetProperties, might be overkill)

Reply:
Still working on this one.

Reply:
Take a look at this. I switched to using enums, though I'm not sure if this is exactly what you were thinking. I didn't go the full factory approach because It seemed like overkill to build a class for each type of energy fraction/multiplicity. It sounds interesting, but I would need help to implement it.

Reply:
Yes, that is what I was thinking. Fine for now